File: dotnet/src/SemanticKernel/CoreSkills/PlannerSkill.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel.Orchestration;
using Microsoft.SemanticKernel.Planning;
using Microsoft.SemanticKernel.SkillDefinition;
using Microsoft.SemanticKernel.Text;

namespace Microsoft.SemanticKernel.CoreSkills;

/// <summary>
/// <para>Semantic skill that creates and executes plans.</para>
/// <para>
/// Usage:
/// var kernel = SemanticKernel.Build(ConsoleLogger.Log);
/// kernel.ImportSkill("planner", new PlannerSkill(kernel));
/// </para>
/// </summary>
public class PlannerSkill
{
    /// <summary>
    /// Parameter names.
    /// <see cref="ContextVariables"/>
    /// </summary>
    public static class Parameters
    {
        /// <summary>
        /// The number of buckets to create
        /// </summary>
        public const string BucketCount = "bucketCount";

        /// <summary>
        /// The prefix to use for the bucket labels
        /// </summary>
        public const string BucketLabelPrefix = "bucketLabelPrefix";

        /// <summary>
        /// The relevancy threshold when filtering registered functions.
        /// </summary>
        public const string RelevancyThreshold = "relevancyThreshold";

        /// <summary>
        /// The maximum number of relevant functions as result of semantic search to include in the plan creation request.
        /// </summary>
        public const string MaxRelevantFunctions = "MaxRelevantFunctions";

        /// <summary>
        /// The list of skills to exclude from the plan creation request.
        /// </summary>
        public const string ExcludedSkills = "excludedSkills";

        /// <summary>
        /// The list of functions to exclude from the plan creation request.
        /// </summary>
        public const string ExcludedFunctions = "excludedFunctions";

        /// <summary>
        /// The list of functions to include in the plan creation request.
        /// </summary>
        public const string IncludedFunctions = "includedFunctions";

        /// <summary>
        /// Whether to use conditional capabilities when creating plans.
        /// </summary>
        public const string UseConditionals = "useConditionals";
    }

    internal sealed class PlannerSkillConfig
    {
        // Depending on the embeddings engine used, the user ask,
        // and the functions available, this value may need to be adjusted.
        // For default, this is set to null to exhibit previous behavior.
        public double? RelevancyThreshold { get; set; }

        // Limits the number of relevant functions as result of semantic
        // search included in the plan creation request.
        // <see cref="Parameters.IncludedFunctions"/> will be included
        // in the plan regardless of this limit.
        public int MaxRelevantFunctions { get; set; } = 100;

        // A list of skills to exclude from the plan creation request.
        public HashSet<string> ExcludedSkills { get; } = new() { RestrictedSkillName };

        // A list of functions to exclude from the plan creation request.
        public HashSet<string> ExcludedFunctions { get; } = new() { "CreatePlan", "ExecutePlan" };

        // A list of functions to include in the plan creation request.
        public HashSet<string> IncludedFunctions { get; } = new() { "BucketOutputs" };

        // Whether to use conditional capabilities when creating plans.
        public bool UseConditionals { get; set; } = false;
    }

    /// <summary>
    /// The name to use when creating semantic functions that are restricted from the PlannerSkill plans
    /// </summary>
    private const string RestrictedSkillName = "PlannerSkill_Excluded";

    /// <summary>
    /// the function flow runner, which executes plans that leverage functions
    /// </summary>
    private readonly FunctionFlowRunner _functionFlowRunner;

    /// <summary>
    /// the bucket semantic function, which takes a list of items and buckets them into a number of buckets
    /// </summary>
    private readonly ISKFunction _bucketFunction;

    /// <summary>
    /// the function flow semantic function, which takes a goal and creates an xml plan that can be executed
    /// </summary>
    private readonly ISKFunction _functionFlowFunction;

    /// <summary>
    /// the conditional function flow semantic function, which takes a goal and creates an xml plan that can be executed
    /// </summary>
    private readonly ISKFunction _conditionalFunctionFlowFunction;

    /// <summary>
    /// Initializes a new instance of the <see cref="PlannerSkill"/> class.
    /// </summary>
    /// <param name="kernel"> The kernel to use </param>
    /// <param name="maxTokens"> The maximum number of tokens to use for the semantic functions </param>
    public PlannerSkill(IKernel kernel, int maxTokens = 1024)
    {
        this._functionFlowRunner = new(kernel);

        this._bucketFunction = kernel.CreateSemanticFunction(
            promptTemplate: SemanticFunctionConstants.BucketFunctionDefinition,
            skillName: RestrictedSkillName,
            maxTokens: maxTokens,
            temperature: 0.0);

        this._functionFlowFunction = kernel.CreateSemanticFunction(
            promptTemplate: SemanticFunctionConstants.FunctionFlowFunctionDefinition,
            skillName: RestrictedSkillName,
            description: "Given a request or command or goal generate a step by step plan to " +
                         "fulfill the request using functions. This ability is also known as decision making and function flow",
            maxTokens: maxTokens,
            temperature: 0.0,
            stopSequences: new[] { "<!--" });

        this._conditionalFunctionFlowFunction = kernel.CreateSemanticFunction(
            promptTemplate: SemanticFunctionConstants.ConditionalFunctionFlowFunctionDefinition,
            skillName: RestrictedSkillName,
            description: "Given a request or command or goal generate a step by step plan to " +
                         "fulfill the request using functions. This ability is also known as decision making and function flow. Uses conditional capabilities",
            maxTokens: maxTokens,
            temperature: 0.0,
            stopSequences: new[] { "<!--" });

        // Currently not exposed -- experimental.
        _ = kernel.CreateSemanticFunction(
            SemanticFunctionConstants.ProblemSolverFunctionDefinition,
            skillName: RestrictedSkillName,
            description: "Given a request or command or goal generate a step by step plan to fulfill the request. " +
                         "This ability is also known as decision making and problem solving",
            maxTokens: maxTokens,
            temperature: 0.0,
            stopSequences: new[] { "<!--" });

        _ = kernel.CreateSemanticFunction(
            SemanticFunctionConstants.SolveNextStepFunctionDefinition,
            skillName: RestrictedSkillName,
            description: "Given a plan with a goal, take the first step and execute it",
            maxTokens: maxTokens,
            temperature: 0.0,
            stopSequences: new[] { "<!-- END -->", "<!--" });
    }

    /// <summary>
    /// When the output of a function is too big, parse the output into a number of buckets.
    /// </summary>
    /// <param name="input"> The input from a function that needs to be parsed into buckets. </param>
    /// <param name="context"> The context to use </param>
    /// <returns> The context with the bucketed results </returns>
    [SKFunction("When the output of a function is too big, parse the output into a number of buckets.")]
    [SKFunctionName("BucketOutputs")]
    [SKFunctionInput(Description = "The output from a function that needs to be parse into buckets.")]
    [SKFunctionContextParameter(Name = Parameters.BucketCount, Description = "The number of buckets.", DefaultValue = "")]
    [SKFunctionContextParameter(
        Name = Parameters.BucketLabelPrefix,
        Description = "The target label prefix for the resulting buckets. " +
                      "Result will have index appended e.g. bucketLabelPrefix='Result' => Result_1, Result_2, Result_3",
        DefaultValue = "Result")]
    public async Task<SKContext> BucketOutputsAsync(string input, SKContext context)
    {
        var bucketsAdded = 0;

        var bucketVariables = new ContextVariables(input);
        if (context.Variables.Get(Parameters.BucketCount, out var bucketCount))
        {
            bucketVariables.Set(Parameters.BucketCount, bucketCount);
        }

        // {'buckets': ['Result 1\nThis is the first result.', 'Result 2\nThis is the second result. It's doubled!', 'Result 3\nThis is the third and final result. Truly astonishing.']}
        var result = await this._bucketFunction.InvokeAsync(new SKContext(bucketVariables, context.Memory, context.Skills, context.Log,
            context.CancellationToken));

        try
        {
            // May need additional formatting here.
            var resultString = result.Result
                .Replace("\\n", "\n")
                .Replace("\n", "\\n");

            var resultObject = JsonSerializer.Deserialize<Dictionary<string, List<string>>>(resultString);

            if (context.Variables.Get(Parameters.BucketLabelPrefix, out var bucketLabelPrefix) &&
                resultObject?.ContainsKey("buckets") == true)
            {
                foreach (var item in resultObject["buckets"])
                {
                    var bucketLabel = $"{bucketLabelPrefix}_{++bucketsAdded}";
                    context.Variables.Set($"{bucketLabel}", item);
                }
            }

            return context;
        }
        catch (Exception e) when (
            e is JsonException or
                InvalidCastException or
                NotSupportedException or
                ArgumentNullException)
        {
            context.Log.LogWarning("Error parsing bucket outputs: {0}", e.Message);
            return context.Fail($"Error parsing bucket outputs: {e.Message}");
        }
    }

    /// <summary>
    /// Create a plan using registered functions to accomplish a goal.
    /// </summary>
    /// <param name="goal"> The goal to accomplish. </param>
    /// <param name="context"> The context to use </param>
    /// <returns> The context with the plan </returns>
    /// <remarks>
    /// The plan is stored in the context as a string. The plan is also stored in the context as a Plan object.
    /// </remarks>
    [SKFunction("Create a plan using registered functions to accomplish a goal.")]
    [SKFunctionName("CreatePlan")]
    [SKFunctionInput(Description = "The goal to accomplish.")]
    [SKFunctionContextParameter(Name = Parameters.RelevancyThreshold, Description = "The relevancy threshold when filtering registered functions.",
        DefaultValue = "")]
    [SKFunctionContextParameter(Name = Parameters.MaxRelevantFunctions,
        Description = "Limits the number of relevant functions as result of semantic search included in the plan creation request.", DefaultValue = "100")]
    [SKFunctionContextParameter(Name = Parameters.ExcludedFunctions, Description = "A list of functions to exclude from the plan creation request.",
        DefaultValue = "")]
    [SKFunctionContextParameter(Name = Parameters.ExcludedSkills, Description = "A list of skills to exclude from the plan creation request.",
        DefaultValue = "")]
    [SKFunctionContextParameter(Name = Parameters.IncludedFunctions, Description = "A list of functions to include in the plan creation request.",
        DefaultValue = "")]
    [SKFunctionContextParameter(Name = Parameters.UseConditionals, Description = "Use conditional functions to create the plan.",
        DefaultValue = "False")]
    public async Task<SKContext> CreatePlanAsync(string goal, SKContext context)
    {
        PlannerSkillConfig config = context.GetPlannerSkillConfig();

        string relevantFunctionsManual = await context.GetFunctionsManualAsync(goal, config);
        context.Variables.Set("available_functions", relevantFunctionsManual);
        // TODO - consider adding the relevancy score for functions added to manual

        var plan = config.UseConditionals
            ? await this._conditionalFunctionFlowFunction.InvokeAsync(context)
            : await this._functionFlowFunction.InvokeAsync(context);

        string fullPlan = $"<{FunctionFlowRunner.GoalTag}>\n{goal}\n</{FunctionFlowRunner.GoalTag}>\n{plan.ToString().Trim()}";
        _ = context.Variables.UpdateWithPlanEntry(new SkillPlan
        {
            Id = Guid.NewGuid().ToString("N"),
            Goal = goal,
            PlanString = fullPlan,
        });

        return context;
    }

    /// <summary>
    /// Execute a plan that uses registered functions to accomplish a goal.
    /// </summary>
    /// <param name="context"> The context to use </param>
    /// <returns> The context with the plan </returns>
    /// <remarks>
    /// The plan is stored in the context as a string. The plan is also stored in the context as a Plan object.
    /// </remarks>
    [SKFunction("Execute a plan that uses registered functions to accomplish a goal.")]
    [SKFunctionName("ExecutePlan")]
    public async Task<SKContext> ExecutePlanAsync(SKContext context)
    {
        var planToExecute = context.Variables.ToPlan();
        try
        {
            var executeResultContext = await this._functionFlowRunner.ExecuteXmlPlanAsync(context, planToExecute.PlanString);
            _ = executeResultContext.Variables.Get(SkillPlan.PlanKey, out var planProgress);
            _ = executeResultContext.Variables.Get(SkillPlan.ResultKey, out var results);

            bool containsPlanTag = planProgress.IndexOf($"<{FunctionFlowRunner.PlanTag}>", StringComparison.OrdinalIgnoreCase) >= 0;
            var isComplete = containsPlanTag && planProgress.IndexOf($"<{FunctionFlowRunner.FunctionTag}", StringComparison.OrdinalIgnoreCase) < 0;
            var isSuccessful = containsPlanTag && !executeResultContext.ErrorOccurred;

            if (string.IsNullOrEmpty(results) && isComplete && isSuccessful)
            {
                results = executeResultContext.Variables.ToString();
            }
            else if (executeResultContext.ErrorOccurred)
            {
                results = executeResultContext.LastErrorDescription;
            }

            _ = context.Variables.UpdateWithPlanEntry(new SkillPlan
            {
                Id = planToExecute.Id,
                Goal = planToExecute.Goal,
                PlanString = planProgress,
                IsComplete = isComplete,
                IsSuccessful = isSuccessful,
                Result = results,
            });

            return context;
        }
        catch (PlanningException e)
        {
            switch (e.ErrorCode)
            {
                case PlanningException.ErrorCodes.InvalidPlan:
                    context.Log.LogWarning("[InvalidPlan] Error executing plan: {0} ({1})", e.Message, e.GetType().Name);
                    _ = context.Variables.UpdateWithPlanEntry(new SkillPlan
                    {
                        Id = Guid.NewGuid().ToString("N"),
                        Goal = planToExecute.Goal,
                        PlanString = planToExecute.PlanString,
                        IsComplete = true, // Plan was invalid, mark complete so it's not attempted further.
                        IsSuccessful = false,
                        Result = e.Message,
                    });

                    return context;
                case PlanningException.ErrorCodes.UnknownError:
                case PlanningException.ErrorCodes.InvalidConfiguration:
                    context.Log.LogWarning("[UnknownError] Error executing plan: {0} ({1})", e.Message, e.GetType().Name);
                    break;
                default:
                    throw;
            }
        }
        catch (Exception e) when (!e.IsCriticalException())
        {
            context.Log.LogWarning("Error executing plan: {0} ({1})", e.Message, e.GetType().Name);
            _ = context.Variables.UpdateWithPlanEntry(new SkillPlan
            {
                Id = Guid.NewGuid().ToString("N"),
                Goal = planToExecute.Goal,
                PlanString = planToExecute.PlanString,
                IsComplete = false,
                IsSuccessful = false,
                Result = e.Message,
            });

            return context;
        }

        return context;
    }
}

File: python/semantic_kernel/core_skills/planner_skill.py
from typing import Set
import json
import uuid
from dataclasses import dataclass

from semantic_kernel.core_skills import constants
from semantic_kernel.kernel_base import KernelBase
from semantic_kernel.kernel import SKContext
from semantic_kernel.kernel_extensions.inline_function_definitions import (
    create_semantic_function,
)
from semantic_kernel.skill_definition import sk_function
from typing import List

from concurrent.futures import ThreadPoolExecutor

from semantic_kernel.kernel import SKContext
from semantic_kernel.memory.null_memory import NullMemory
from semantic_kernel.skill_definition.function_view import FunctionView
from semantic_kernel.memory.memory_query_result import MemoryQueryResult
from semantic_kernel.planning.function_flow_runner import FunctionFlowRunner
from semantic_kernel.planning.plan import Plan


@dataclass
class Parameters:
    bucket_count = "bucketCount"
    bucket_label_prefix = "bucketLabelPrefix"
    relevancy_threshold = "relevancyThreshold"
    max_relevant_functions = "MaxRelevantFunctions"
    excluded_skills = "excludedSkills"
    excluded_functions = "excludedFunctions"
    included_functions = "includedFunctions"


class PlannerSkillConfig:
    def __init__(self):
        self.relevancy_threshold: float = None
        self.max_relevant_functions: int = 100
        self.excluded_skills: Set[str] = set([PlannerSkill.RESTRICTED_SKILL_NAME])
        self.excluded_functions: Set[str] = set(["CreatePlan", "ExecutePlan"])
        self.included_functions: Set[str] = set(["BucketOutputs"])
        self.included_functions = "includedFunctions"


class SKContextPlanning:
    PlannerMemoryCollectionName = "Planning.SKFunctionsManual"
    PlanSKFunctionsAreRemembered = "Planning.SKFunctionsAreRemembered"

    @staticmethod
    async def get_functions_manual_async(
        context: SKContext,
        semantic_query: str = None,
        config: PlannerSkillConfig = None,
    ) -> str:
        config = config or PlannerSkillConfig()
        functions = await SKContextPlanning.get_available_functions_async(
            context, config, semantic_query
        )

        return "\n\n".join(x.to_manual_string() for x in functions)

    @staticmethod
    async def get_available_functions_async(
        context: SKContext,
        config: PlannerSkillConfig = None,
        semantic_query: str = None,
    ) -> List[FunctionView]:
        excluded_skills = config.excluded_skills or []
        excluded_functions = config.excluded_functions or []
        included_functions = config.included_functions or []

        context.throw_if_skill_collection_not_set()
        functions_view = context.skills.get_functions_view()

        available_functions = [
            f for x in functions_view._semantic_functions.values() for f in x
        ] + [f for x in functions_view._native_functions.values() for f in x]

        available_functions = [
            f
            for f in available_functions
            if f.skill_name not in excluded_skills and f.name not in excluded_functions
        ]

        if (
            not semantic_query
            or isinstance(context.memory, NullMemory)
            or config.relevancy_threshold is None
        ):
            # If no semantic query is provided, return all available functions.
            # If a Memory provider has not been registered, return all available functions.
            result = available_functions
        else:
            result = []
            # Remember functions in memory so that they can be searched.
            await SKContextPlanning.remember_functions_async(
                context, available_functions
            )

            # Search for functions that match the semantic query.
            memories = context.memory.search_async(
                SKContextPlanning.PlannerMemoryCollectionName,
                semantic_query,
                config.max_relevant_functions,
                config.relevancy_threshold,
                context._cancellation_token,
            )

            # Add functions that were found in the search results.
            result += await SKContextPlanning.get_relevant_functions_async(
                context, available_functions, memories
            )

            # Add any missing functions that were included but not found in the search results.
            missing_functions = [
                f for f in included_functions if f.name not in [x.name for x in result]
            ]
            missing_functions = [
                af
                for f in missing_functions
                for af in available_functions
                if f == af.name
            ]

            result += missing_functions

        return result

    @staticmethod
    async def get_relevant_functions_async(
        context: SKContext,
        available_functions: List[FunctionView],
        memories: List[MemoryQueryResult],
    ) -> List[FunctionView]:
        relevant_functions = []
        with ThreadPoolExecutor() as executor:
            for memory_entry in memories:
                function = next(
                    (
                        x
                        for x in available_functions
                        if x.to_fully_qualified_name() == memory_entry.id
                    ),
                    None,
                )
                if function:
                    context.log.debug(
                        "Found relevant function. Relevance Score: %s, Function: %s",
                        memory_entry.relevance,
                        function.to_fully_qualified_name(),
                    )
                    relevant_functions.append(function)

        return relevant_functions

    @staticmethod
    async def remember_functions_async(
        context: SKContext, available_functions: List[FunctionView]
    ) -> None:
        # Check if the functions have already been saved to memory.
        if context.variables.get(SKContextPlanning.PlanSKFunctionsAreRemembered):
            return

        for function in available_functions:
            function_name = function.to_fully_qualified_name()
            key = function.description if function.description else function_name

            # It'd be nice if there were a saveIfNotExists method on the memory interface
            memory_entry = await context.memory.get_async(
                SKContextPlanning.PlannerMemoryCollectionName,
                key,
                context._cancellation_token,
            )
            if not memory_entry:
                # TODO It'd be nice if the min_relevance_score could be a parameter for each item that was saved to memory
                # As folks may want to tune their functions to be more or less relevant.
                await context.memory.save_information_async(
                    SKContextPlanning.PlannerMemoryCollectionName,
                    key,
                    function_name,
                    function.to_manual_string(),
                    context._cancellation_token,
                )

        # Set a flag to indicate that the functions have been saved to memory.
        context.variables.set(SKContextPlanning.PlanSKFunctionsAreRemembered, "true")

    @staticmethod
    def get_planner_skill_config(context: SKContext) -> PlannerSkillConfig:
        config = PlannerSkillConfig()

        threshold = context.variables.get(Parameters.relevancy_threshold)
        if threshold is not None:
            parsed_value = float(threshold)
            config.relevancy_threshold = parsed_value

        max_relevant_functions = context.variables.get(
            Parameters.max_relevant_functions
        )
        if max_relevant_functions is not None:
            parsed_value = int(max_relevant_functions)
            config.max_relevant_functions = parsed_value

        excluded_functions = context.variables.get(Parameters.excluded_functions)
        if excluded_functions is not None:
            excluded_functions_list = excluded_functions.split(",").map(str.strip)
            # Excluded functions and excluded skills from context.variables should be additive to the default excluded functions and skills.
            config.excluded_functions.update(
                excluded_functions_list - config.excluded_functions
            )

        excluded_skills = context.variables.get(Parameters.excluded_skills)
        if excluded_skills is not None:
            excluded_skills_list = excluded_skills.split(",").map(str.strip)
            # Excluded functions and excluded skills from context.variables should be additive to the default excluded functions and skills.
            config.excluded_skills.update(excluded_skills_list - config.excluded_skills)

        included_functions = context.variables.get(Parameters.included_functions)
        if included_functions is not None:
            included_functions_list = included_functions.split(",").map(str.strip)
            # Included functions from context.variables should not override the default excluded functions.
            config.included_functions.update(
                included_functions_list - config.excluded_functions
            )
            config.included_functions.difference_update(config.excluded_functions)

        return config


class PlannerSkill(SKContextPlanning):
    RESTRICTED_SKILL_NAME = "PlannerSkill_Excluded"

    def __init__(
        self,
        kernel: KernelBase,
        max_tokens: int = 1024,
    ):
        super().__init__()
        # TODO: implement function flow runner
        # self._functionFlowRunner = FunctionFlowRunner(kernel)

        # TODO: implement bucket function
        # self._bucketFunction = create_semantic_function(
        #     kernel=kernel,
        #     prompt_template=constants.BucketFunctionDefinition,
        #     skill_name=self.RESTRICTED_SKILL_NAME,
        #     maxTokens=maxTokens,
        #     temperature=0.0,
        # )
        self._function_flow_runner = FunctionFlowRunner(kernel)

        self._function_flow_function = create_semantic_function(
            kernel=kernel,
            skill_name=self.RESTRICTED_SKILL_NAME,
            prompt_template=constants.FUNCTION_FLOW_FUNCTION_DEFINITION,
            description="Given a request or command or goal generate a step by step plan to "
            + "fulfill the request using functions. This ability is also known as decision making and function flow",
            max_tokens=max_tokens,
            temperature=0.0,
            stop_sequences=["<!--"],
        )

    @staticmethod
    def get_planner_skill_config(context: SKContext) -> PlannerSkillConfig:
        config = PlannerSkillConfig()

        if context.variables.get(Parameters.relevancy_threshold)[0]:
            config.relevancy_threshold = context.variables.get(
                Parameters.relevancy_threshold
            )[1]

        if context.variables.get(Parameters.max_relevant_functions)[0]:
            config.max_relevant_functions = context.variables.get(
                Parameters.max_relevant_functions
            )[1]

        if context.variables.get(Parameters.excluded_functions)[0]:
            excluded_functions = context.variables.get(Parameters.excluded_functions[1])
            config.excluded_functions = set(
                [x.strip() for x in excluded_functions.split(",")]
            )

        if context.variables.get(Parameters.excluded_skills)[0]:
            excluded_skills = context.variables.get(Parameters.excluded_skills[1])
            config.excluded_skills = set(
                [x.strip() for x in excluded_skills.split(",")]
            )

        if context.variables.get(Parameters.included_functions)[0]:
            included_functions = context.variables.get(Parameters.included_functions)[1]
            included_functions_set = set(
                [x.strip() for x in included_functions.split(",")]
            )
            # Included functions from context.Variables should not override the default excluded functions.
            config.included_functions |= (
                included_functions_set - config.excluded_functions
            )
            config.included_functions -= config.excluded_functions

        return config

    @staticmethod
    def update_context_with_plan_entry(plan: Plan, context: SKContext) -> None:
        context.variables.update(plan.to_json())
        context.variables.set(plan.ID_KEY, plan.id)
        context.variables.set(plan.GOAL_KEY, plan.goal)
        context.variables.set(plan.PLAN_KEY, plan.plan_string)
        context.variables.set(plan.IS_COMPLETE_KEY, plan.is_complete)
        context.variables.set(plan.IS_SUCCESSFUL_KEY, plan.is_successful)
        context.variables.set(plan.RESULT_KEY, plan.result)

    @sk_function(
        name="CreatePlan",
        description="Create a plan using registered functions to accomplish a goal.",
    )
    async def create_plan_async(self, goal: str, context: SKContext) -> str:
        config = self.get_planner_skill_config(context=context)

        relevant_functions_manual = await self.get_functions_manual_async(
            semantic_query=goal, context=context, config=config
        )

        context.variables.set("available_functions", relevant_functions_manual)

        plan = await self._function_flow_function.invoke_async(context=context)

        full_plan = f"<{self._function_flow_runner.GOAL_TAG}>{goal}</{self._function_flow_runner.GOAL_TAG}>\n{plan.result.strip()}"

        _ = PlannerSkill.update_context_with_plan_entry(
            plan=Plan(id=str(uuid.uuid4()), goal=goal, plan_string=full_plan),
            context=context,
        )

        return context

    @sk_function(
        name="ExecutePlan",
        description="Execute a plan using registered functions to accomplish a goal.",
    )
    async def execute_plan_async(self, context: SKContext) -> SKContext:
        plan_to_execute = Plan.from_json(context.variables)
